topics/ICRC-21/ICRC-21.did

```diff
@@ -0,0 +1,142 @@
+        LineDisplay: record {
+            // Maximum number of characters that can be displayed per line.
+            characters_per_line: nat;
+            // Maximum number of lines that can be displayed at once on a single page.
+            lines_per_page: nat;
+        };
+    };
+};
+
+type icrc21_consent_message_request = record {
+    // Method name of the canister call.
+    method: text;
+    // Argument of the canister call.
+    arg: blob;
+    // User preferences with regards to the consent message presented to the end-user.
+    user_preferences: icrc21_consent_message_spec;
+};
+
+type icrc21_consent_message = variant {
+    // Message for a generic display able to handle large documents and do proper line wrapping and pagination / scrolling.
+    // Uses Markdown formatting, no external resources (e.g. images) are allowed.
+    GenericDisplayMessage: text;
+    // Message for a simple display able to handle pages with multiple lines of text with a fixed maximum number of
+    // characters per line.
+    // Multiple pages can be used if the text does no fit on a single page.
+    // Uses plain text, without any embedded formatting elements.
+    LineDisplayMessage: record {
+        pages: vec record {
+            // Lines of text to be displayed on a single page.
+            // Must not have more entries (lines) than specified in the icrc21_consent_message_spec.
+            // Lines must not exceed the number of characters per line specified in the icrc21_consent_message_spec.
+            lines: vec text;
+        };
+    };
+};
+
+type icrc21_consent_info = record {
+    // Consent message describing in a human-readable format what the call will do.
+    //
+    // The message should adhere as close as possible to the user_preferences specified in the consent_message_spec
+    // of the icrc21_consent_message_request.
+    // If the message is not available for the given user_preferences a fallback message may be used.
+    // The consent_message_spec must match the consent_message provided.
+    //
+    // The message should be short and concise.
+    // It should only contain information that is:
+    // * relevant to the user
+    // * relevant given the canister call argument
+    //
+    // The message must fit the following context shown to
+    // the user on the signer UI:
+    // ┌─────────────────────────────────┐
+    // │  Approve the following action?  │
+    // │  ┌───────────────────────────┐  │
+    // │  │    <consent_message>      │  │
+    // │  └───────────────────────────┘  │
+    // │  ┌───────────┐   ┌───────────┐  │
+    // │  │  Reject   │   │  Approve  │  │
+    // │  └───────────┘   └───────────┘  │
+    // └─────────────────────────────────┘
+    consent_message: icrc21_consent_message;
+    // Actual consent message spec used to generate the message.
+    // Clients must reject the consent message if the spec does not match the supplied consent_message.
```

Comment:
nit: this implies that the client must recognise  e.g. whether the message is in the specified language, which might be difficult in practice.  Or is the rejection here meant to be "best effort"?

Reply:
@przydatek: I have now bundled non-enforced `metadata` from the `consent_message_spec`. Now the `device_spec` is no longer included in the response and hence cannot cause a conflict with the `icrc21_consent_message` variant, thus sidestepping the issue.

Please take another look.